### PR TITLE
Document issue with Kokkos::sort and comparators

### PR DESCRIPTION
**Pull request by masterleinad**
Related to https://github.com/kokkos/kokkos/pull/6801#issuecomment-1969888881.

### CODE REVIEW

File: docs/source/known-issues.rst [hunk 1]
```diff
@@ -23,6 +23,16 @@ HIP
...
 
+SYCL
+====
+
+- Multiple of the Kokkos algorithm functions use third-party libraries like oneDPL.
+  When using these, Kokkos doesn't control the kernel launch and thus the user has to make sure that all arguments
+  that are forwarded to the TPL satisfy the sycl::is_device_copyable trait to avoid compiler errors. This holds true in particular
+  for comparators used with Kokkos::sort in Kokkos versions prior to 4.3. The best advice to give is to make sure the respective
+  parameters are trivially-copyable. If this isn't possible, sycl::is_device_copyable should be specialized and users should make
+  sure to use raw pointers instead of Kokkos::Views.
```
Comment (dalg24) [inline docs/source/known-issues.rst:34]: Can you add a code example and include the compile error so that it can be easily found by users?

Comment (dalg24) [reply docs/source/known-issues.rst:34]:
```C++
Kokkos::sort(exec, values, KOKKOS_LAMBDA(int i, int j) { return keys(i) < keys(j); });  // error blah with SYCL

Comment (masterleinad) [reply docs/source/known-issues.rst:34]: 25284b5d

Comment (masterleinad) [reply docs/source/known-issues.rst:34]: 25284b5d

File: docs/source/known-issues.rst [hunk 2]
```diff
@@ -23,6 +23,49 @@ HIP
 
   gcc 7, 9, and later do not have this issue.
 
+SYCL
+====
+
+- Multiple of the Kokkos algorithm functions use third-party libraries like oneDPL.
```
Comment (dalg24) [inline docs/source/known-issues.rst:29]: "Multiple of" sounds weird to me.
Maybe "several" or "a number of"?  Maybe a native English speaker can advise here.

File: docs/source/known-issues.rst [hunk 3]
```diff
@@ -23,6 +23,49 @@ HIP
...
+      1694 |     unpack<KernelName, KernelType, PropertiesT,
+           |     ^
+     /usr/bin/compiler/../../include/sycl/handler.hpp:1293:7: note: in instantiation of function template specialization 'sycl::handler::kernel_parallel_for_wrapper<sycl::detail::RoundedRangeKernel<sycl::item<1, true>, 1, (lambda at /usr/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1578:83)>, sycl::item<1, true>, sycl::detail::RoundedRangeKernel<sycl::item<1, true>, 1, (lambda at /usr/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:1578:83)>, sycl::ext::oneapi::experimental::properties<std::tuple<>>>' requested here
+      1293 |       kernel_parallel_for_wrapper<KName, TransformedArgType, decltype(Wrapper),
+           |       ^
+     /usr/bin/compiler/../../include/sycl/handler.hpp:2332:5: note: (skipping 7 contexts in backtrace; use -ftemplate-backtrace-limit=0 to see all)
+      2332 |     parallel_for_lambda_impl<KernelName, KernelType, 1, PropertiesT>(
+           |     ^
+     [...]
```
Comment (dalg24) [inline docs/source/known-issues.rst:67]: Please add a Functor definition followed by a `sycl::is_device_copyable` traits specialization to showcase what the user is expected to do.

File: docs/source/known-issues.rst (hunk 2 above)
Comment (masterleinad) [reply docs/source/known-issues.rst]: 4e016a2ccc4ecd462a87709762e6009f0327ae91

File: docs/source/known-issues.rst (hunk 3 above)
Comment (masterleinad) [reply docs/source/known-issues.rst:68]: 5f4bda354aee113e32358df35a81802bd3666eee

Comment (dalg24) [reply docs/source/known-issues.rst:68]: I meant also adding a definition for `MyComparator` and showing the actual `Kokkos::sort` invocation.
Do you think it is too obvious?

Comment (dalg24) [reply docs/source/known-issues.rst:68]: Oh I thought you'd transform the lambda into that comparator thing.
I suppose that is good enough.

Comment (dalg24) [reply docs/source/known-issues.rst:68]: The error you posted was with a lambda though right?  (Not necessarily asking you to update)

Comment (masterleinad) [reply docs/source/known-issues.rst:68]: Yes, it was with a lambda but I printed only those errors that are independent of the example. The lambda referenced in it comes from `oneDPL`.

**Review by dalg24 — APPROVED** [review]

File: docs/source/known-issues.rst [hunk 4]
```diff
@@ -23,6 +23,58 @@ HIP
...
   gcc 7, 9, and later do not have this issue.
 
+SYCL
+====
+
+- Several of the Kokkos algorithm functions use third-party libraries like oneDPL.
+  When using these, Kokkos doesn't control the kernel launch and thus the user has to make sure that all arguments
+  that are forwarded to the TPL satisfy the sycl::is_device_copyable trait to avoid compiler errors. This holds true in particular
+  for comparators used with Kokkos::sort in Kokkos versions prior to 4.3. The best advice to give is to make sure the respective
+  parameters are trivially-copyable. If this isn't possible, sycl::is_device_copyable should be specialized and users should make
```
Comment (science-enthusiast) [inline docs/source/known-issues.rst:33]: For Kokkos release 4.7, could we also say the following?

oneDPL version should be >= 2022.8.0, apart from specializing `sycl::is_device_copyable`.

We could also say the following: It is enough if the user defines a device copyable comparator. Kokkos internally defines the specialization of `sycl::is_device_copyable` for that user defined comparator.

**Pull request merged**